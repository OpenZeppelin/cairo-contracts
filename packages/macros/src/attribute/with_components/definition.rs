use cairo_lang_formatter::format_string;
use cairo_lang_macro::{attribute_macro, Diagnostic, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use regex::Regex;

use super::{components::AllowedComponents, parser::WithComponentsParser};

/// Inserts multiple component dependencies into a modules codebase.
#[attribute_macro]
pub fn with_components(attribute_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    let args = parse_args(&attribute_stream.to_string());

    // 1. Get the components info (if valid)
    let mut components_info = vec![];
    let empty_result = ProcMacroResult::new(TokenStream::empty());
    for arg in args {
        let maybe_component = AllowedComponents::from_str(&arg);
        match maybe_component {
            Ok(component) => {
                components_info.push(component.get_info());
            }
            Err(err) => {
                return empty_result.with_diagnostics(err.into());
            }
        }
    }

    // 2. Parse the item stream
    let db = SimpleParserDatabase::default();
    let (content, diagnostics) = match db.parse_virtual(item_stream.to_string()) {
        Ok(node) => WithComponentsParser::new(node, &components_info).parse(&db),
        Err(err) => {
            let error = Diagnostic::error(err.format(&db));
            return empty_result.with_diagnostics(error.into());
        }
    };

    let formatted_content = if !content.is_empty() {
        format_string(&db, content)
    } else {
        content
    };

    ProcMacroResult::new(TokenStream::new(formatted_content)).with_diagnostics(diagnostics)
}

/// Parses the arguments from the attribute stream.
fn parse_args(text: &str) -> Vec<String> {
    let re = Regex::new(r"(\w+)").unwrap();
    let matches = re.find_iter(text);
    matches.map(|m| m.as_str().to_string()).collect()
}

#[cfg(test)]
mod tests {
    use super::parse_args;

    #[test]
    fn test_parse_args() {
        let attribute = "(ERC20, Ownable)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20", "Ownable"]);

        let attribute = "ERC20";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20"]);

        let attribute = "(Ownable, ERC20, Other, Another)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["Ownable", "ERC20", "Other", "Another"]);
    }
}
