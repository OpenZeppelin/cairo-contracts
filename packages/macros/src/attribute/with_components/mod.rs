pub mod components;
pub mod definition;
pub mod diagnostics;
pub mod parser;
