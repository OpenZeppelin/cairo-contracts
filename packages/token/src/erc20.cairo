pub mod erc20;
pub mod interface;
pub mod snip12_utils;

pub use erc20::{ERC20Component, ERC20HooksEmptyImpl};
pub use interface::{ERC20ABIDispatcher, ERC20ABIDispatcherTrait};
