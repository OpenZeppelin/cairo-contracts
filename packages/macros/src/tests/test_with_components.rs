use crate::with_components::with_components_avevetedp5blk as with_components;
use cairo_lang_macro::TokenStream;
use indoc::{formatdoc, indoc};
use insta::assert_snapshot;

#[test]
fn test_with_erc20() {
    let attribute = "(ERC20)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc20.initializer(\"MyToken\", \"MTK\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc20_no_initializer() {
    let attribute = "(ERC20)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc20_no_hooks_impl() {
    let attribute = "(ERC20)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc20.initializer(\"MyToken\", \"MTK\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_ownable() {
    let attribute = "(Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod Owned {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, owner: ContractAddress) {
                self.ownable.initializer(owner);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_ownable_no_initializer() {
    let attribute = "(Ownable)";
    let item = indoc!(
        "
      #[starknet::contract]
      pub mod Owned {
          use starknet::ContractAddress;

          #[storage]
          pub struct Storage {}

          #[constructor]
          fn constructor(ref self: ContractState) {
          }
      }
      "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_two_components() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, owner: ContractAddress) {
                self.ownable.initializer(owner);
                self.erc20.initializer(\"MyToken\", \"MTK\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_two_components_no_initializer() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_two_components_no_constructor() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_access_control() {
    let attribute = "(AccessControl)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod Contract {
            use openzeppelin_access::accesscontrol::DEFAULT_ADMIN_ROLE;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, default_admin: ContractAddress) {
                self.accesscontrol.initializer();

                self.accesscontrol._grant_role(DEFAULT_ADMIN_ROLE, default_admin);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_access_control_no_initializer() {
    let attribute = "(AccessControl)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod Contract {
            use openzeppelin_access::accesscontrol::DEFAULT_ADMIN_ROLE;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_vesting() {
    let attribute = "(Vesting, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod VestingWallet {
            use openzeppelin_finance::vesting::LinearVestingSchedule;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(
                ref self: ContractState,
                beneficiary: ContractAddress,
                start: u64,
                duration: u64,
                cliff_duration: u64,
            ) {
                self.ownable.initializer(beneficiary);
                self.vesting.initializer(start, duration, cliff_duration);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_vesting_no_initializer() {
    let attribute = "(Vesting, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod VestingWallet {
            use openzeppelin_finance::vesting::LinearVestingSchedule;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(
                ref self: ContractState,
                beneficiary: ContractAddress,
                start: u64,
                duration: u64,
                cliff_duration: u64,
            ) {
                self.ownable.initializer(beneficiary);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_vesting_no_schedule() {
    let attribute = "(Vesting, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod VestingWallet {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(
                ref self: ContractState,
                beneficiary: ContractAddress,
                start: u64,
                duration: u64,
                cliff_duration: u64,
            ) {
                self.ownable.initializer(beneficiary);
                self.vesting.initializer(start, duration, cliff_duration);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_no_contract_attribute() {
    let attribute = "(Ownable)";
    let item = indoc!(
        "
        pub mod Owned {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, owner: ContractAddress) {
                self.ownable.initializer(owner);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_no_body() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        pub mod MyContract;
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_no_components() {
    let attribute = "()";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_invalid_component() {
    let attribute = "(ERC6000, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_src5() {
    let attribute = "(SRC5)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_initializable() {
    let attribute = "(Initializable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[generate_trait]
            #[abi(per_item)]
            impl ExternalImpl of ExternalTrait {
                #[external(v0)]
                fn initialize(ref self: ContractState) {
                    self.initializer.initialize();
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_initializable_no_initialize_call() {
    let attribute = "(Initializable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_pausable() {
    let attribute = "(Pausable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[generate_trait]
            #[abi(per_item)]
            impl ExternalImpl of ExternalTrait {
                #[external(v0)]
                fn pause(ref self: ContractState) {
                    self.pausable.pause();
                }

                #[external(v0)]
                fn unpause(ref self: ContractState) {
                    self.pausable.unpause();
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_pausable_no_pause_call() {
    let attribute = "(Pausable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[generate_trait]
            #[abi(per_item)]
            impl ExternalImpl of ExternalTrait {
                #[external(v0)]
                fn unpause(ref self: ContractState) {
                    self.pausable.unpause();
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_pausable_no_unpause_call() {
    let attribute = "(Pausable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[generate_trait]
            #[abi(per_item)]
            impl ExternalImpl of ExternalTrait {
                #[external(v0)]
                fn pause(ref self: ContractState) {
                    self.pausable.pause();
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_pausable_no_pause_or_unpause_call() {
    let attribute = "(Pausable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_reentrancy_guard() {
    let attribute = "(ReentrancyGuard)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721() {
    let attribute = "(ERC721)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_token::erc721::ERC721HooksEmptyImpl;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc721.initializer(\"MyToken\", \"MTK\", \"\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721_no_initializer() {
    let attribute = "(ERC721)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_token::erc721::ERC721HooksEmptyImpl;

            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721_no_hooks_impl() {
    let attribute = "(ERC721)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc721.initializer(\"MyToken\", \"MTK\", \"\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc1155() {
    let attribute = "(ERC1155)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_token::erc1155::ERC1155HooksEmptyImpl;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc1155.initializer(\"\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc1155_no_initializer() {
    let attribute = "(ERC1155)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_token::erc1155::ERC1155HooksEmptyImpl;

            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc1155_no_hooks_impl() {
    let attribute = "(ERC1155)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc1155.initializer(\"\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721_enumerable() {
    let attribute = "(ERC721Enumerable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc721_enumerable.initializer();
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721_enumerable_no_initializer() {
    let attribute = "(ERC721Enumerable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721_receiver() {
    let attribute = "(ERC721Receiver)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc721_receiver.initializer();
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc721_receiver_no_initializer() {
    let attribute = "(ERC721Receiver)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc1155_receiver() {
    let attribute = "(ERC1155Receiver)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc1155_receiver.initializer();
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc1155_receiver_no_initializer() {
    let attribute = "(ERC1155Receiver)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc2981() {
    let attribute = "(ERC2981)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_token::common::erc2981::DefaultConfig;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, default_royalty_receiver: ContractAddress) {
                self.erc2981.initializer(default_royalty_receiver, 0);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc2981_no_initializer() {
    let attribute = "(ERC2981)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_token::common::erc2981::DefaultConfig;

            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc2981_no_config() {
    let attribute = "(ERC2981)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, default_royalty_receiver: ContractAddress) {
                self.erc2981.initializer(default_royalty_receiver, 0);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc2981_no_initializer_no_config() {
    let attribute = "(ERC2981)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, default_royalty_receiver: ContractAddress) {
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_upgradeable() {
    let attribute = "(Upgradeable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[abi(embed_v0)]
            impl UpgradeableImpl of IUpgradeable<ContractState> {
                fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
                    self.upgradeable.upgrade(new_class_hash);
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_upgradeable_no_upgrade_call() {
    let attribute = "(Upgradeable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}

            #[abi(embed_v0)]
            impl UpgradeableImpl of IUpgradeable<ContractState> {
                fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_nonces() {
    let attribute = "(Nonces)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_multisig() {
    let attribute = "(Multisig)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, quorum: u32, signers: Span<ContractAddress>) {
                self.multisig.initializer(quorum, signers);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_multisig_no_initializer() {
    let attribute = "(Multisig)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_timelock_controller() {
    let attribute = "(TimelockController)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(
                ref self: ContractState,
                min_delay: u64,
                proposers: Span<ContractAddress>,
                executors: Span<ContractAddress>,
                admin: ContractAddress,
            ) {
                self.timelock_controller.initializer(min_delay, proposers, executors, admin);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_timelock_controller_no_initializer() {
    let attribute = "(TimelockController)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_votes() {
    let attribute = "(Votes)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            use openzeppelin_utils::cryptography::snip12::SNIP12Metadata;

            #[storage]
            pub struct Storage {}

            /// Required for hash computation.
            pub impl SNIP12MetadataImpl of SNIP12Metadata {
                fn name() -> felt252 {
                    'DAPP_NAME'
                }
                fn version() -> felt252 {
                    'DAPP_VERSION'
                }
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_votes_no_metadata() {
    let attribute = "(Votes)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyContract {
            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

//
// Helpers
//

/// Returns a string representation of the result of the macro expansion,
/// including the token stream, diagnostics and aux data.
fn get_string_result(attribute: &str, item: &str) -> String {
    let attribute_stream = TokenStream::new(attribute.to_string());
    let item_stream = TokenStream::new(item.to_string());
    let raw_result = with_components(attribute_stream, item_stream);
    let none = "None";

    let mut token_stream = raw_result.token_stream.to_string();
    let mut diagnostics = String::new();
    for d in raw_result.diagnostics {
        diagnostics += format!("====\n{:?}: {}====", d.severity, d.message).as_str();
    }

    if token_stream.is_empty() {
        token_stream = none.to_string();
    }
    if diagnostics.is_empty() {
        diagnostics = none.to_string();
    }

    formatdoc! {
        "
        TokenStream:

        {}

        Diagnostics:

        {}

        AuxData:

        {:?}
        ",
        token_stream, diagnostics, raw_result.aux_data
    }
}
