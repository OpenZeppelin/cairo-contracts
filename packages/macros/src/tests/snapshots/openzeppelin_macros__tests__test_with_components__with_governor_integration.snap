---
source: src/tests/test_with_components.rs
expression: result
snapshot_kind: text
---
TokenStream:

#[starknet::contract]
pub mod GovernorTimelockedMock {
    use openzeppelin_governance::governor::DefaultConfig;
    use openzeppelin_utils::cryptography::snip12::SNIP12Metadata;
    use starknet::ContractAddress;

    pub const VOTING_DELAY: u64 = 86400; // 1 day
    pub const VOTING_PERIOD: u64 = 604800; // 1 week
    pub const PROPOSAL_THRESHOLD: u256 = 10;
    pub const QUORUM: u256 = 100_000_000;

    // Governor
    #[abi(embed_v0)]
    impl GovernorImpl = GovernorComponent::GovernorImpl<ContractState>;

    // Extensions external
    #[abi(embed_v0)]
    impl VotesTokenImpl = GovernorVotesComponent::VotesTokenImpl<ContractState>;
    #[abi(embed_v0)]
    impl GovernorSettingsAdminImpl =
        GovernorSettingsComponent::GovernorSettingsAdminImpl<ContractState>;
    #[abi(embed_v0)]
    impl TimelockedImpl =
        GovernorTimelockExecutionComponent::TimelockedImpl<ContractState>;

    // SRC5
    #[abi(embed_v0)]
    impl SRC5Impl = SRC5Component::SRC5Impl<ContractState>;

    #[storage]
    struct Storage {
        #[substorage(v0)]
        pub governor: GovernorComponent::Storage,
        #[substorage(v0)]
        pub governor_votes: GovernorVotesComponent::Storage,
        #[substorage(v0)]
        pub governor_settings: GovernorSettingsComponent::Storage,
        #[substorage(v0)]
        pub governor_counting_simple: GovernorCountingSimpleComponent::Storage,
        #[substorage(v0)]
        pub governor_timelock_execution: GovernorTimelockExecutionComponent::Storage,
        #[substorage(v0)]
        pub src5: SRC5Component::Storage,
    }

    #[constructor]
    fn constructor(
        ref self: ContractState, votes_token: ContractAddress, timelock_controller: ContractAddress,
    ) {
        self.governor.initializer();
        self.governor_votes.initializer(votes_token);
        self.governor_settings.initializer(VOTING_DELAY, VOTING_PERIOD, PROPOSAL_THRESHOLD);
        self.governor_timelock_execution.initializer(timelock_controller);
    }

    //
    // SNIP12 Metadata
    //

    pub impl SNIP12MetadataImpl of SNIP12Metadata {
        fn name() -> felt252 {
            'DAPP_NAME'
        }

        fn version() -> felt252 {
            'DAPP_VERSION'
        }
    }

    //
    // Locally implemented extensions
    //

    impl GovernorQuorum of GovernorComponent::GovernorQuorumTrait<ContractState> {
        /// See `GovernorComponent::GovernorQuorumTrait::quorum`.
        fn quorum(self: @GovernorComponent::ComponentState<ContractState>, timepoint: u64) -> u256 {
            QUORUM
        }
    }
    use openzeppelin_governance::governor::GovernorComponent;
    use openzeppelin_governance::governor::extensions::{
        GovernorCountingSimpleComponent, GovernorSettingsComponent,
        GovernorTimelockExecutionComponent, GovernorVotesComponent,
    };
    use openzeppelin_introspection::src5::SRC5Component;

    component!(path: GovernorComponent, storage: governor, event: GovernorEvent);
    component!(path: GovernorVotesComponent, storage: governor_votes, event: GovernorVotesEvent);
    component!(
        path: GovernorSettingsComponent, storage: governor_settings, event: GovernorSettingsEvent,
    );
    component!(
        path: GovernorCountingSimpleComponent,
        storage: governor_counting_simple,
        event: GovernorCountingSimpleEvent,
    );
    component!(
        path: GovernorTimelockExecutionComponent,
        storage: governor_timelock_execution,
        event: GovernorTimelockExecutionEvent,
    );
    component!(path: SRC5Component, storage: src5, event: SRC5Event);

    impl GovernorInternalImpl = GovernorComponent::InternalImpl<ContractState>;
    impl GovernorVotesInternalImpl = GovernorVotesComponent::InternalImpl<ContractState>;
    impl GovernorVotesGovernorVotes = GovernorVotesComponent::GovernorVotes<ContractState>;
    impl GovernorSettingsInternalImpl = GovernorSettingsComponent::InternalImpl<ContractState>;
    impl GovernorSettingsGovernorSettings =
        GovernorSettingsComponent::GovernorSettings<ContractState>;
    impl GovernorCountingSimpleGovernorCounting =
        GovernorCountingSimpleComponent::GovernorCounting<ContractState>;
    impl GovernorTimelockExecutionInternalImpl =
        GovernorTimelockExecutionComponent::InternalImpl<ContractState>;
    impl GovernorTimelockExecutionGovernorExecution =
        GovernorTimelockExecutionComponent::GovernorExecution<ContractState>;
    impl SRC5InternalImpl = SRC5Component::InternalImpl<ContractState>;

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        #[flat]
        GovernorEvent: GovernorComponent::Event,
        #[flat]
        GovernorVotesEvent: GovernorVotesComponent::Event,
        #[flat]
        GovernorSettingsEvent: GovernorSettingsComponent::Event,
        #[flat]
        GovernorCountingSimpleEvent: GovernorCountingSimpleComponent::Event,
        #[flat]
        GovernorTimelockExecutionEvent: GovernorTimelockExecutionComponent::Event,
        #[flat]
        SRC5Event: SRC5Component::Event,
    }
}


Diagnostics:

None

AuxData:

None
